### ISSUE TITLE
Deprecate support for SQL queries as part of data sources

### ISSUE THREAD

**Issue by woop** [feast-dev/feast#1688]
In the current Feast API it is possible for users to specify either a `table reference` or `query` as part of many data sources 
* [BigQuery](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/data_source.py#L130)
* [Redshift](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/data_source.py#L927)

The motivation for supporting `query` was to allow users to manipulate source data prior to reading it into Feast during materialization or training dataset building. The assumption is that not all users own their sources, so they may not be able to make a change to the table schema. Or they may not have permissions to create views in their offline store.

However, we are unsure whether this assumption holds up. It seems like most users only need
1. Field mapping
2. Column projection
3. Filtering of the source rows

Supporting `query` comes with major downsides
1. It's wasteful. Feast has to execute the full query in order to do simple operations like get metadata (column names/types)
2. Feast is unable to optimize the underlying query and is forced to execute it as-is, sometimes repeatedly. This can lead to increased costs for users.
3. It requires the Feast team to maintain two separate code paths for functionality that is almost identical.

All of the required functionality above can be added as configuration options as part of `table ref`, which would free us up from having to support two separate means of querying offline data. I'm hoping to hear from users whether continuing to support `query` is important, or whether we can try to support similar functionality with `table ref`.

**Comment by MattDelac** [comment]
This would be great and we can definitely optimized a lot down the road 👍

**Comment by shihgianlee** [comment]
It makes sense to support `table ref` only and won't impact us if we got rid of `query`. 

Our BigQuery (BQ) tables are pretty straightforward since the data transformation logic is taken care of by external process before populating the BigQuery. We want to keep as little logic as possible in BiqQuery or Feast so that we don't have to migrate them if we ever decided to switch product. We create BQ view to do simple BQ tables join to expose data to Feast.

As users, we may not have permissions to create BQ views. But, DevOps can help with that. If the data is external source, we have a process to ingest those data into our BQ tables. With limited information, I don't understand the BQ view permissions problem faced by the users. In my experience, that has never been a problem for me.

**Comment by woop** [comment]
> bles are pretty straightforward since the data transformation logic is taken care of by external process before populating the BigQu

Great feedback @shihgianlee!

**Comment by hugoleodev** [comment]
Hi guys, in my case that I will expose you below.

```python
stock_source = BigQuerySource(
    query="SELECT TIMESTAMP(`Date`) as event_timestamp, * "
    "FROM `cool_project.happy_dataset.stocks`"
)
```
 The column Date type is 'DATE' and I need to map to  'TIMESTAMP'.

How can I do it with column mapping?

**Comment by woop** [comment]
> Hi guys, in my case that I will expose you below.
> 
> ```python
> stock_source = BigQuerySource(
>     query="SELECT TIMESTAMP(`Date`) as event_timestamp, * "
>     "FROM `cool_project.happy_dataset.stocks`"
> )
> ```
> 
> The column Date type is 'DATE' and I need to map to 'TIMESTAMP'.
> 
> How can I do it with column mapping?

Good point. In your case you need type casting. I'm leaning towards asking you to create a BigQuery view. Would that be reasonable?

**Comment by mavysavydav** [comment]
what's the difference between field mapping and column projection?

**Comment by mavysavydav** [comment]
This change is cool with twitter

**Comment by mavysavydav** [comment]
To confirm. -- Would the 3 pieces of functionality (Field mapping, Column projection, Filtering of the source rows) be supported before this `query` is deprecated?

**Comment by hugoleodev** [comment]
> > Hi guys, in my case that I will expose you below.
> > ```python
> > stock_source = BigQuerySource(
> >     query="SELECT TIMESTAMP(`Date`) as event_timestamp, * "
> >     "FROM `cool_project.happy_dataset.stocks`"
> > )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The column Date type is 'DATE' and I need to map to 'TIMESTAMP'.
> > How can I do it with column mapping?
> 
> Good point. In your case you need type casting. I'm leaning towards asking you to create a BigQuery view. Would that be reasonable?

Yes, it sounds reasonable for me.

**Comment by woop** [comment]
> To confirm -- Would the 3 pieces of functionality (Field mapping, Column projection, Filtering of the source rows) be supported before this `query` is deprecated?

Yea. Why don't we try to be really explicit on the scope.

- [x] Field mapping
- [ ] Projection (being able to select a subset of columns)
- [ ] Filtering of rows

I think type casting should be out of scope for now. I think the real question is how filtering of rows should work. Is it a predicate like `my_column !='blah' AND other_column>0`, or something more abstracted? @mavysavydav do you have an opinion on what you'd like to see there, or is it not really something you'd need?

**Comment by mavysavydav** [comment]
are the checkboxes supposed to convey that field mapping is already supported but we're looking to possibly support the other 2? just making sure, i do recall there being support for field mapping.

**Comment by woop** [comment]
Yea we have field mappings in sources already.
